homeassistant/components/reolink/views.py

```diff
@@ -85,63 +87,75 @@ async def get(
 
         if _LOGGER.isEnabledFor(logging.DEBUG):
             _LOGGER.debug(
-                "Opening VOD stream from %s: %s",
+                "Opening proxy VOD stream from %s: %s",
                 host.api.camera_name(ch),
                 host.api.hide_password(reolink_url),
             )
 
+        await self._download_mutex.acquire()
         try:
-            reolink_response = await self.session.get(
-                reolink_url,
-                timeout=ClientTimeout(
-                    connect=15, sock_connect=15, sock_read=5, total=None
-                ),
-            )
-        except ClientError as err:
-            err_str = host.api.hide_password(
-                f"Reolink playback error while getting mp4: {err!s}"
-            )
-            if retry <= 0:
-                _LOGGER.warning(err_str)
+            try:
+                reolink_response = await self.session.get(
+                    reolink_url,
+                    timeout=ClientTimeout(
+                        connect=15, sock_connect=15, sock_read=5, total=None
+                    ),
+                )
+            except ClientError as err:
+                err_str = host.api.hide_password(
+                    f"Reolink playback error while getting mp4: {err!s}"
+                )
+                if retry <= 0:
+                    _LOGGER.warning(err_str)
+                    return web.Response(body=err_str, status=HTTPStatus.BAD_REQUEST)
+                _LOGGER.debug("%s, renewing token", err_str)
+                await host.api.expire_session(unsubscribe=False)
+                self._download_mutex.release()
```

Comment:
Why are you realising the lock here manually?

Reply:
Because if you don't release the lock here, the next call `await self.get(.....` which is a retry of this whole function will get stuck in a dead-lock since it will try to do `await self._download_mutex.acquire()` while it is already acquired (one function up in the callstack, which was the first try).

Reply:
Can we create a new function where we use the lock? So we could handle the exception and call the new function in this function so we don't have the recursive call.
I don't like to manually handle the acquiring/releasing of locks as it could lead to errors. For example, in the future, we have another use case where we need to recursively call this function. In this case, the dev needs to know that he needs to manually release the lock first, which is weird, imo.